text/3516-public-private-dependencies.md

```diff
@@ -0,0 +1,425 @@
+manifest and rustc will warn saying that `serde` and `serde_json` are private
+dependencies in a public API.
+In 2024+ editions, this will be an error.
+
+To resolve the warning in a semver compatible way, they would need to declare both dependencies as public:
+```toml
+[package]
+name = "diagnostic"
+version = "1.0.0"
+
+[dependencies]
+serde = { version = "1", features = ["derive"], public = true }
+serde_json = { version = "1", public = true }
+```
+For edition migrations, `cargo fix` will look for the warning code and mark those dependencies as `public`.
+
+However, for this example, it was an oversight in exposing `serde_json` in the public API.
+Note that removing it from the public API is a semver incompatible change.
+```toml
+[package]
+name = "diagnostic"
+version = "1.0.0"
+
+[dependencies]
+serde = { version = "1", features = ["derive"], public = true }
+serde_json = "1"
+```
+```rust
+#[derive(Clone, Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+pub struct Diagnostic {
+    code: String,
+    message: String,
+    file: std::path::PathBuf,
+    span: std::ops::Range<usize>,
+}
+
+impl std::str::FromStr for Diagnostic {
+    type Err = Error
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        serde_json::from_str(s).map_err(Error)
+    }
+}
+
+pub struct Error(serde_json::Error);
+```
+
+If you then had a public dependency on `diagnostic`,
+then `serde` would automatically be considered a public dependency of yours.
+
+At times, some public dependencies are effectively private.
+Take this code from older versions of `clap`
+```rust
+#[doc(hidden)]
+#[cfg(feature = "derive")]
+pub mod __derive_refs {
+    #[doc(hidden)]
+    pub use once_cell;
+}
+```
+Since the proc-macro can only guarantee that the namespace `clap` is accessible,
+`clap` must re-export any functionality that is needed at runtime by the generated code.
+As a last-ditch way of dealing with this, a user may allow the error:
+```rust
+#[doc(hidden)]
+#[allow(external_private_dependency)]
+#[cfg(feature = "derive")]
+pub mod __derive_refs {
+    #[doc(hidden)]
+    pub use once_cell;
+}
+```
+A similar case is pub-in-private:
+```rust
+mod private {
+    #[allow(external_private_dependency)]
+    pub struct Foo { pub x: some_dependency::SomeType }
+}
+```
+Though this might be worked around by reducing the visibility to `pub(crate)`.
+
+I say "last ditch" because in most other cases,
+a user would be better served by wrapping the API which would be helped with
+features like `impl Trait` in type aliases if we had it.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## rustc
+
+The main change to the compiler will be to accept a new modifier on the `--extern` flag that Cargo
+supplies which marks it as a private dependency.
+The modifier will be called `priv` (e.g. `--extern priv:serde`).
+The compiler then emits the lint `external_private_dependency` if it encounters private
```

Comment:
Is this suggesting to rename the existing lint to `external_private_dependency` or is this a different lint from [`exported_private_dependencies`](https://doc.rust-lang.org/nightly/rustc/lints/listing/warn-by-default.html#exported-private-dependencies)?

For background information: It was decided during the implementation phase to change the lint name from the original RFC 1977 text. Should this all be updated to the new name?

Implementation-wise, are there any changes needed on the rustc-side?

Reply:
Thanks!  I tried to update everything for the current implementation and missed that one.

> Implementation-wise, are there any changes needed on the rustc-side?

I expect this section is the only change:

> `exported-private-dependencies` will be `allow` by default for pre-2024 editions.
It will be a member of the `rust-2024-compatibility` lint group so that it gets automatically picked up by `cargo fix --edition`.
In the 2024 edition, this lint will be `deny`.